File: pkg/services/dashboard_manager.go
package services

import (
	"fmt"
	"sync"
	"time"
)

// DashboardManager 仪表板管理器
// 管理控制台输出和状态显示
type DashboardManager struct {
	mu sync.RWMutex
}

// BotStatus 机器人状态
type BotStatus struct {
	Name          string
	Phase         string
	PriceYes      float64
	PriceNo       float64
	YesShares     float64
	YesAvg        float64
	NoShares      float64
	NoAvg         float64
	ActiveOrders  int
	PnL           float64
	UnrealizedPnL float64
	IsComplete    bool
}

var globalDashboard = &DashboardManager{}

// GetDashboard 获取全局仪表板实例
func GetDashboard() *DashboardManager {
	return globalDashboard
}

// Update 更新仪表板显示
func (dm *DashboardManager) Update(status BotStatus) {
	dm.mu.Lock()
	defer dm.mu.Unlock()

	fmt.Printf("\n")
	fmt.Printf("═══════════════════════════════════════════════════════════════\n")
	fmt.Printf("  %s - %s\n", status.Name, status.Phase)
	fmt.Printf("═══════════════════════════════════════════════════════════════\n")
	fmt.Printf("  价格: YES $%.3f | NO $%.3f\n", status.PriceYes, status.PriceNo)
	fmt.Printf("  持仓: YES %.2f@$%.3f | NO %.2f@$%.3f\n",
		status.YesShares, status.YesAvg, status.NoShares, status.NoAvg)
	fmt.Printf("  活跃订单: %d\n", status.ActiveOrders)
	
	if status.UnrealizedPnL != 0 {
		pnlColor := "🟢"
		if status.UnrealizedPnL < 0 {
			pnlColor = "🔴"
		}
		fmt.Printf("  %s 未实现盈亏: $%.4f\n", pnlColor, status.UnrealizedPnL)
	}
	
	if status.IsComplete && status.PnL != 0 {
		pnlColor := "🟢"
		if status.PnL < 0 {
			pnlColor = "🔴"
		}
		fmt.Printf("  %s 实现盈亏: $%.4f\n", pnlColor, status.PnL)
	}
	fmt.Printf("═══════════════════════════════════════════════════════════════\n\n")
}

// Log 记录消息
func (dm *DashboardManager) Log(message, level string) {
	dm.mu.Lock()
	defer dm.mu.Unlock()

	timestamp := time.Now().Format("15:04:05")
	icon := "ℹ️"
	
	switch level {
	case "INFO":
		icon = "ℹ️"
	case "WARN":
		icon = "⚠️"
	case "ERROR":
		icon = "❌"
	case "TRADE":
		icon = "💰"
	}

	fmt.Printf("[%s] %s %s\n", timestamp, icon, message)
}

// LogEvent 记录事件
func (dm *DashboardManager) LogEvent(strategy, message string) {
	dm.mu.Lock()
	defer dm.mu.Unlock()

	timestamp := time.Now().Format("15:04:05")
	fmt.Printf("[%s] [%s] %s\n", timestamp, strategy, message)
}

// SetBotStatus 设置机器人状态
func SetBotStatus(name, status string) {
	fmt.Printf("🤖 [%s] %s\n", name, status)
}

File: _upstream/src/services/DashboardManager.ts
/**
 * Centralized Dashboard Manager
 * "Black Box" Logger - Persistent Error History
 * Nothing gets deleted, all errors are logged permanently
 */

import { FileLogger } from "./FileLogger";

export interface BotState {
  name: string;
  phase: string;
  priceYes: number;
  priceNo: number;
  yesShares: number;
  yesAvg: number;
  noShares: number;
  noAvg: number;
  activeOrders: number;
  pnl?: number;
  hedgePrice?: number;
  hedgeSide?: string;
  isComplete: boolean;
  extraInfo?: string;
  status?: string;
  unrealizedPnL?: number; // For stop loss display
}

// Global status for when bots are between cycles
const botStatus: Map<string, string> = new Map();

export function setBotStatus(name: string, status: string) {
  botStatus.set(name, status);
  DashboardManager.log(`STATUS CHANGE: ${status}`, "INFO");
}

export function getBotStatus(name: string): string | undefined {
  return botStatus.get(name);
}

class DashboardManagerClass {
  private lastStatus: any = {};

  // 🟢 PERSISTENT HISTORY (The "Black Box")
  // We keep the last 50 lines so you can scroll back and see exactly what happened.
  private eventHistory: string[] = [];

  // 🟢 UNIFIED LOGGING (Info, Trade, Warn, Error)
  public log(
    message: string,
    type: "INFO" | "TRADE" | "WARN" | "ERROR" = "INFO"
  ) {
    const time = new Date().toLocaleTimeString();
    let icon = "🔹";

    if (type === "TRADE") icon = "💥";
    if (type === "WARN") icon = "⚠️";
    if (type === "ERROR") icon = "❌";

    const logLine = `[${time}] ${icon} ${message}`;

    // Add to history
    this.eventHistory.push(logLine);
    // Keep last 50 lines (increased from 20 for deep debugging)
    if (this.eventHistory.length > 50) this.eventHistory.shift();

    // 🟢 BLACK BOX: Save to File
    FileLogger.log(type, message);

    // Force render immediately so you see it
    this.render();
  }

  // --- WRAPPERS FOR COMPATIBILITY ---
  public logEvent(source: string, message: string) {
    // Hook this one too - it will call log() which saves to file
    this.log(`[${source}] ${message}`, "TRADE");
  }

  public logError(source: string, message: string) {
    this.log(`${source}: ${message}`, "ERROR");
  }

  public logSystem(message: string) {
    this.log(message, "INFO");
  }

  public logSystemEvent(message: string) {
    this.log(message, "INFO");
  }

  public update(status: any) {
    this.lastStatus = status;
    this.render();
  }

  private render() {
    console.clear();
    const s = this.lastStatus;

    // 1. STATUS HEADER
    console.log(
      `╔════════════════════════════════════════════════════════════════════════════╗`
    );
    console.log(
      `║ 🤖 POLYMARKET WAR ROOM (${new Date().toLocaleTimeString()})             ║`
    );
    console.log(
      `╠════════════════════════════════════════════════════════════════════════════╣`
    );
    console.log(`║  🔫 PHASE: ${(s.phase || "Waiting").padEnd(67)} ║`);
    console.log(
      `║  📊 YES: $${(s.priceYes?.toFixed(2) || "0.00").padEnd(6)} | NO: $${(
        s.priceNo?.toFixed(2) || "0.00"
      ).padEnd(6)}                             ║`
    );
    console.log(
      `║  💰 PnL: $${(s.pnl?.toFixed(2) || "0.00").padEnd(6)} | 🛡️ Hedge: ${(
        s.hedgeSide || "-"
      ).padEnd(3)} @ $${(s.hedgePrice?.toFixed(2) || "0.00").padEnd(
        6
      )}            ║`
    );
    console.log(
      `║  📉 STOP LOSS CHECK: ${(s.unrealizedPnL
        ? "$" + s.unrealizedPnL.toFixed(2)
        : "---"
      ).padEnd(58)} ║`
    );
    console.log(
      `╠════════════════════════════════════════════════════════════════════════════╣`
    );

    // 2. THE BLACK BOX (HISTORY)
    console.log(
      `║ 📜 EVENT LOG (ERRORS & TRADES)                                             ║`
    );
    console.log(
      `╠────────────────────────────────────────────────────────────────────────────╣`
    );

    if (this.eventHistory.length === 0) {
      console.log(
        `║ (Waiting for events...)                                                    ║`
      );
    } else {
      this.eventHistory.forEach((line) => {
        // Cut string to fit box width
        const safeMsg = line.substring(0, 74).padEnd(74);
        console.log(`║ ${safeMsg} ║`);
      });
    }
    console.log(
      `╚════════════════════════════════════════════════════════════════════════════╝`
    );
  }
}

// Singleton instance
export const DashboardManager = new DashboardManagerClass();
